### PR TITLE
contrib/raftexample: fix restart path

### LINKED ISSUE #4474: raftexample restart without leader
single raft cluster. First start can vote a leader, but restart cannot vote leader. It always follower

### PR DESCRIPTION
**Pull request by heyitsanthony**
The ConfChange fix crashes WAL replay because it assumed the node
always exists. Now, only ApplyConfChange if node is non-nil since
the raft node restart logic handles the conf changes in the log.

Also changed restart to only RestartNode if there's more than one
peer in the cluster; otherwise the node deadlocks waiting for an
election. This fixes #4474

### CODE REVIEW

**Comment by xiang90** [comment]
>  otherwise the node deadlocks waiting for an election. 

I think for the one node case, the election should be successful since the node will vote for itself.

File: contrib/raftexample/raft.go [hunk 1]
```diff
@@ -103,8 +103,9 @@ func (rc *raftNode) publishEntries(ents []raftpb.Entry) bool {
 		case raftpb.EntryConfChange:
 			var cc raftpb.ConfChange
 			cc.Unmarshal(ents[i].Data)
-
-			rc.node.ApplyConfChange(cc)
+			if rc.node != nil {
```
Comment (xiang90) [inline contrib/raftexample/raft.go]: instead of doing this, we probably should replay log after we start/restart the node?

**Comment by heyitsanthony** [comment]
On init it gives:
```
raft2016/02/09 20:59:11 INFO: 1 became follower at term 0
raft2016/02/09 20:59:11 INFO: newRaft 1 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2016/02/09 20:59:11 INFO: 1 became follower at term 1
raft2016/02/09 20:59:12 INFO: 1 is starting a new election at term 1
raft2016/02/09 20:59:12 INFO: 1 became candidate at term 2
raft2016/02/09 20:59:12 INFO: 1 received vote from 1 at term 2
raft2016/02/09 20:59:12 INFO: 1 became leader at term 2
raft2016/02/09 20:59:12 INFO: raft.node: 1 elected leader 1 at term 2
```

On restart it gives:
```
raft2016/02/09 21:07:17 INFO: 1 became follower at term 0
raft2016/02/09 21:07:17 INFO: newRaft 1 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 3, lastterm: 2]
```

So it's not starting an election for some reason...

File: contrib/raftexample/raft.go (hunk 1 above)
Comment (heyitsanthony) [inline contrib/raftexample/raft.go]: Tried that first, it crashes:
```
raft2016/02/09 21:11:36 INFO: newRaft 1 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2016/02/09 21:11:37 INFO: 1 is starting a new election at term 0
raft2016/02/09 21:11:37 INFO: 1 became candidate at term 1
raft2016/02/09 21:11:37 INFO: 1 received vote from 1 at term 1
raft2016/02/09 21:11:37 unstable.slice[1,4) out of bound [1,1]
panic: unstable.slice[1,4) out of bound [1,1]
```

**Comment by xiang90** [comment]
@heyitsanthony I will give it a look.

**Comment by heyitsanthony** [comment]
fixed up ptal /cc @xiang90

File: contrib/raftexample/raft.go (hunk 1 above)
Comment (xiang90) [inline contrib/raftexample/raft.go]: we do not need this check anymore, right?

**Comment by xiang90** [comment]
LGTM

File: contrib/raftexample/raft.go (hunk 1 above)
Comment (heyitsanthony) [inline contrib/raftexample/raft.go]: right, nuked

**Comment by heyitsanthony** [comment]
ok merging on greenlight

**Pull request merged**